src/dataflow-types/src/reconciliation/command.rs

```diff
@@ -0,0 +1,361 @@
+                        let iter = frontier.update_iter(changes.drain());
+                        changes.extend(iter);
+                    } else {
+                        changes.clear();
+                    }
+                }
+
+                self.responses
+                    .push_back(Response::Storage(StorageResponse::TimestampBindings(
+                        feedback,
+                    )));
+            }
+            Response::Compute(ComputeResponse::PeekResponse(uuid, response), instance) => {
+                self.responses.push_back(Response::Compute(
+                    ComputeResponse::PeekResponse(uuid, response),
+                    instance,
+                ));
+            }
+            Response::Compute(ComputeResponse::TailResponse(id, response), instance) => {
+                self.responses.push_back(Response::Compute(
+                    ComputeResponse::TailResponse(id, response),
+                    instance,
+                ));
+            }
+        }
+    }
+
+    async fn observe_compute_command(
+        &mut self,
+        command: ComputeCommand,
+        instance: ComputeInstanceId,
+    ) -> Result<(), anyhow::Error> {
+        use Command::Compute;
+        use ComputeCommand::*;
+        match command {
+            CreateInstance(config, logging) => {
+                if !self.created {
+                    self.created = true;
+                    if let Some(logging) = &logging {
+                        for id in logging.log_identifiers() {
+                            if !self.uppers.contains_key(&(id, Some(instance))) {
+                                self.start_tracking(id, Some(instance));
+                            }
+                        }
+                    }
+                    self.client
+                        .send(Compute(CreateInstance(config, logging), instance))
+                        .await?;
+                }
+                Ok(())
+            }
+            cmd @ DropInstance => {
+                if self.created {
+                    self.created = false;
+                    self.uppers.clear();
+                    self.client.send(Compute(cmd, instance)).await
+                } else {
+                    Ok(())
+                }
+            }
+            CreateDataflows(dataflows) => {
+                let mut create = Vec::new();
+                for dataflow in dataflows {
+                    let mut create_dataflow = false;
+                    for id in dataflow
+                        .sink_exports
+                        .iter()
+                        .map(|(id, _)| *id)
+                        .chain(dataflow.index_exports.iter().map(|(id, _, _)| *id))
+                    {
+                        if !self.uppers.contains_key(&(id, Some(instance))) {
+                            create_dataflow = true;
+                            self.start_tracking(id, Some(instance));
+                        }
+                    }
```

Comment:
This is a bit wonky, but fair enough, that if the new command is "more than" a previous dataflow, we'll ship instructions that rebind an existing id. I wonder if we might prefer to just retain the commands that installed things (e.g. dataflows, by `dataflow_id`) and check that they are exact.

Reply:
Reworked the logic. It now `assert_eq` that the newly installed dataflow matches the existing one. Deferring to whoever owns the plan and the guarantees around it!